File: 蓝桥杯/第8届/个人赛省赛/C语言B组/4.cpp
/*
标题：方格分割

6x6的方格，沿着格子的边线剪开成两部分。
要求这两部分的形状完全相同。

如图：p1.png, p2.png, p3.png 都是可行的分割法。

试计算：
包括这3种分法在内，一共有多少种不同的分割方法。
注意：旋转对称的属于同一种分割法。

请提交该整数，不要填写任何多余的内容或说明文字。
*/

/*
 * 解题思路：
 * 6×6 方格有 7×7 个格点。剪开线必须经过中心点 (3,3)（保证两部分对称）。
 * 从中心 (3,3) 出发 DFS 走到边界，每走一步同时标记对称点 (6-x, 6-y)。
 * 若当前点 (x,y) 的对称点已标记，说明剪开线闭合，形成两个相同部分。
 *
 * 但要求两部分形状完全相同，即剪开线必须是中心对称的。
 * DFS 路径本身即为中心对称：每走一步，对称地标记对立点。
 * 走到边界时形成一条完整的分割线。
 *
 * 四种旋转被视为同一种，故答案需除以 4。
 * 已知答案：509。
 */
#include <cstring>
#include <iostream>
using namespace std;

const int N = 6;
bool vis[N + 1][N + 1];
int ans = 0;
int dir[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

void dfs(int x, int y) {
    if (x == 0 || x == N || y == 0 || y == N) {
        ans++;
        return;
    }
    vis[x][y] = true;
    vis[N - x][N - y] = true;
    for (int i = 0; i < 4; i++) {
        int nx = x + dir[i][0];
        int ny = y + dir[i][1];
        if (nx >= 0 && nx <= N && ny >= 0 && ny <= N && !vis[nx][ny])
            dfs(nx, ny);
    }
    vis[x][y] = false;
    vis[N - x][N - y] = false;
}

int main() {
    memset(vis, 0, sizeof(vis));
    dfs(N / 2, N / 2);
    // 旋转对称（4种旋转视为同一种），且从中心出发的4个方向起始相同
    cout << ans / 4 << endl;
    return 0;
}

File: 蓝桥杯/第8届/个人赛省赛/Java语言A组/_4.java
/*
标题：方格分割

6x6的方格，沿着格子的边线剪开成两部分。
要求这两部分的形状完全相同。

如图：p1.png, p2.png, p3.png 就是可行的分割法。

试计算：
包括这3种分法在内，一共有多少种不同的分割方法。
注意：旋转对称的属于同一种分割法。

请提交该整数，不要填写任何多余的内容或说明文字。


解题思路：
从中心点 (3,3) 出发，DFS 沿着格线对称裁剪。
方格 6×6 的格线交点坐标范围为 0~6。
中心对称条件：若某条边被剪开，则关于中心 (3,3) 对称的边也必须被剪开。
DFS 时同时标记当前点和对称点，保证两部分连通且分割线连续。
到达边界时找到一种分割方法。
由于每种分割被计 4 次（起点的 4 个方向），结果除以 4。
答案：509
*/
public class _4 {
    static int[] dx = {0, 1, 0, -1};
    static int[] dy = {1, 0, -1, 0};
    static boolean[][] vis = new boolean[7][7];
    static int ans = 0;

    public static void main(String[] args) {
        vis[3][3] = true;
        dfs(3, 3);
        System.out.println(ans / 4);
    }

    static void dfs(int x, int y) {
        if (x == 0 || x == 6 || y == 0 || y == 6) {
            ans++;
            return;
        }
        for (int k = 0; k < 4; k++) {
            int nx = x + dx[k], ny = y + dy[k];
            int sx = 6 - nx, sy = 6 - ny; // 对称点
            if (!vis[nx][ny]) {
                vis[nx][ny] = vis[sx][sy] = true;
                dfs(nx, ny);
                vis[nx][ny] = vis[sx][sy] = false;
            }
        }
    }
}
